### PR TITLE
add ecc encrypt&decrypt

### LINKED ISSUE #2594: Add ECC encryption and decryption tool method to core
**Summary or problem description**
The key distribution process of the random number on the chain requires asymmetric encryption of the key(https://github.com/neo-project/neo-modules/issues/637#issue-981193765). At the same time, a data encryption function can be added to the Fs client. So we can add an asymmetric encryption and decryption tool method to the core Cryptography, easy to use.

**Do you have any solution you want to propose?**

**Process**

**Encrypt**

- Generate a symmetric encryption private key
- Use the symmetric encryption private key to encrypt data
- Use the recipient's public key to encrypt the symmetric encryption private key
- Send the encrypted symmetric encryption private key and data to the receiver

**Decrypt**

- The receiver receives the data and obtains the encrypted symmetric encryption private key and data from it
- Use your own private key to decrypt the encrypted symmetric encryption private key
- Use symmetric encryption private key to decrypt encrypted data

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Cryptography

If you think it is possible, I can complete the relevant PR.

### PR DESCRIPTION
**Pull request by doubiliu**
Close #2594

### CODE REVIEW

File: src/neo/Cryptography/Helper.cs [hunk 1]
```diff
@@ -109,6 +111,88 @@ public static byte[] Sha256(this Span<byte> value)
...
+            }
+            byte[] RBar = R.EncodePoint(true);
+            var z = ECPoint.Multiply(pubKey, r).X; // z = r * P = r* k * G
+            var Z = z.ToByteArray();
+            var EK = Z.Sha256();
+            var EM = message.AES256Encrypt(EK);
+            return RBar.Concat(EM).ToArray();
+        }
+        public static byte[] ECDecrypt(byte[] cypher, byte[] priKey, ECPoint pubKey)
+        {
```
Comment (shargon) [inline src/neo/Cryptography/Helper.cs:164]: We should check that the pubKey is not infinite.

File: src/neo/Cryptography/Helper.cs [hunk 2]
```diff
@@ -109,6 +111,88 @@ public static byte[] Sha256(this Span<byte> value)
...
+                while (rx.Sign == 0);
+            }
+            byte[] RBar = R.EncodePoint(true);
+            var z = ECPoint.Multiply(pubKey, r).X; // z = r * P = r* k * G
+            var Z = z.ToByteArray();
+            var EK = Z.Sha256();
+            var EM = message.AES256Encrypt(EK);
+            return RBar.Concat(EM).ToArray();
+        }
+        public static byte[] ECDecrypt(byte[] cypher, byte[] priKey, ECPoint pubKey)
```
Comment (shargon) [inline src/neo/Cryptography/Helper.cs:163]: `KeyPair` instead of private and public?

Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs:163]: OK.I will change it.

File: src/neo/Cryptography/Helper.cs (hunk 1 above)
Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs:164]: why?

Comment (shargon) [reply src/neo/Cryptography/Helper.cs:164]: We should validate always all the inputs

Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs:164]: I have check in ecrypt

Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs:165]: check X == null && Y == null?

File: src/neo/Cryptography/Helper.cs [hunk 3]
```diff
@@ -109,6 +112,90 @@ public static byte[] Sha256(this Span<byte> value)
...
+            aes.Padding = PaddingMode.PKCS7;
+            using ICryptoTransform decryptor = aes.CreateDecryptor();
+            return decryptor.TransformFinalBlock(data, 0, data.Length);
+        }
+
+        public static byte[] AES256Encrypt(this byte[] data, byte[] key)
+        {
+            using Aes aes = Aes.Create();
+            aes.Key = key;
+            aes.Mode = CipherMode.ECB;
```
[the remark this answers is not in the document]
Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs]: I have changed

[the remark this answers is not in the document]
Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs]: @roman-khimov

**Comment by Liaojinghui** [comment]
Test cases crash the `dotnet` on my Macbook.

File: src/neo/Cryptography/Helper.cs [hunk 4]
```diff
@@ -109,6 +115,104 @@ public static byte[] Sha256(this Span<byte> value)
...
+                    do
+                    {
+                        r = rng.NextBigInteger((int)curve.N.GetBitLength());
+                    }
+                    while (r.Sign == 0 || r.CompareTo(curve.N) >= 0);
+                    R = ECPoint.Multiply(curve.G, r);
+                    BigInteger x = R.X.Value;
+                    rx = x.Mod(curve.N);
+                }
+                while (rx.Sign == 0);
```
Comment (Liaojinghui) [inline src/neo/Cryptography/Helper.cs:170]:
```suggestion
                } while (rx.Sign == 0);
```

File: src/neo/Cryptography/Helper.cs [hunk 5]
```diff
@@ -109,6 +115,104 @@ public static byte[] Sha256(this Span<byte> value)
...
+            }
+            byte[] RBar = R.EncodePoint(true);
+            var z = ECPoint.Multiply(pubKey, r).X; // z = r * P = r* k * G
+            var Z = z.ToByteArray();
+            var EK = Z.Sha256();
```
Comment (Liaojinghui) [inline src/neo/Cryptography/Helper.cs:175]:
```suggestion
            var EK = ECPoint.Multiply(pubKey, r).X.ToByteArray().Sha256(); // z = r * P = r* k * G
```

**Comment by ZhangTao1596** [comment]
> Test cases crash the `dotnet` on my Macbook.
> 
> Its this code that cause the crash:
> 
> ```cs
> using var cipher = new AesGcm(keyBytes);
> ```

`AesGcm` depends on  `openssl@1.1`. But there is `LibreSSL` as default in maxOS. 
1. install openssl `brew install openssl@1/1`
2. Add `export DYLD_LIBRARY_PATH="/usr/local/opt/openssl@1.1/lib"` in env variables
3.  Use `dotnet test --filter "FullyQualifiedName=Neo.UnitTests.Cryptography.UT_Cryptography_Helper.TestECEncryptAndDecrypt"` run the test.

**Comment by Liaojinghui** [comment]
@ZhangTao1596  Thank you, will try it out

File: src/neo/Cryptography/Helper.cs [hunk 6]
```diff
@@ -1,13 +1,19 @@
 using Neo.IO;
 using Neo.Network.P2P.Payloads;
+using Neo.Wallets;
+using Org.BouncyCastle.Crypto.Engines;
+using Org.BouncyCastle.Crypto.Modes;
+using Org.BouncyCastle.Crypto.Parameters;
```
Comment (Liaojinghui) [inline src/neo/Cryptography/Helper.cs:6]:
```suggestion
```

**Review by Liaojinghui — DISMISSED** [review]

File: src/neo/Cryptography/Helper.cs [hunk 7]
```diff
@@ -109,6 +115,91 @@ public static byte[] Sha256(this Span<byte> value)
...
+            if (nonce.Length != 12) throw new ArgumentException();
+            var cipherBytes = new byte[plainData.Length];
+            var tag = new byte[16];
+            using var cipher = new AesGcm(key);
+            cipher.Encrypt(nonce, plainData, cipherBytes, tag, associatedData);
+            var cipherWithTag = new byte[nonce.Length + cipherBytes.Length + tag.Length];
+            Buffer.BlockCopy(nonce, 0, cipherWithTag, 0, nonce.Length);
+            Buffer.BlockCopy(cipherBytes, 0, cipherWithTag, nonce.Length, cipherBytes.Length);
+            Buffer.BlockCopy(tag, 0, cipherWithTag, nonce.Length + cipherBytes.Length, tag.Length);
+            return cipherWithTag;
```
Comment (Liaojinghui) [inline src/neo/Cryptography/Helper.cs:130]:
```suggestion
            var keyLen = key == null ? 0 : key.Length;
            var nonceLen = nonce == null ? 0 : nonce.Length;

            if (keyLen != 32) throw new ArgumentException();
            if (nonceLen != 12) throw new ArgumentException();

            var msgLen = plainData == null ? 0 : plainData.Length;
            var tagLen = 16;

            var cipherBytes = new byte[msgLen];
            var tag = new byte[tagLen];
            using var cipher = new AesGcm(key);
            cipher.Encrypt(nonce, plainData, cipherBytes, tag, associatedData);
            var cipherWithTag = new byte[nonceLen + msgLen + tagLen];
            Buffer.BlockCopy(nonce, 0, cipherWithTag, 0, nonceLen);
            Buffer.BlockCopy(cipherBytes, 0, cipherWithTag, nonceLen, msgLen);
            Buffer.BlockCopy(tag, 0, cipherWithTag, nonceLen + msgLen, tagLen);
            return cipherWithTag;
```

Comment (Liaojinghui) [reply src/neo/Cryptography/Helper.cs:130]: non-functional suggestions, personal preference, you don't have to follow them.

**Review by Liaojinghui — DISMISSED** [review]

**Review by nicolegys — APPROVED** [review]

File: src/neo/Cryptography/Helper.cs [hunk 8]
```diff
@@ -156,62 +154,57 @@ public static byte[] AES256Decrypt(this byte[] encryptedData, byte[] key, byte[]
             return decryptedData;
         }
 
-        public static byte[] ECEncrypt(byte[] message, ECPoint pubKey)
+        public static byte[] EcdhEncrypt(byte[] message, KeyPair local, ECPoint remote)
```
Comment (Liaojinghui) [inline src/neo/Cryptography/Helper.cs:157]: There is no such `ecdhencrypt` and `ecdhdecrypt` term. You only need one single function to derive the ecdh shared key, the ecrypt and decrypt would be that of AES.

Comment (Liaojinghui) [reply src/neo/Cryptography/Helper.cs:157]:
```C#
        public static byte[] ECDHDeriveKey(KeyPair local, ECPoint remote)
        {
            ECDiffieHellman ecdh1 = ECDiffieHellmanCng.Create(new ECParameters
            {
                Curve = ECCurve.NamedCurves.nistP256,
                D = local.PrivateKey,
                Q = new System.Security.Cryptography.ECPoint
                {
                    X = local.PublicKey.EncodePoint(false)[1..][..32],
                    Y = local.PublicKey.EncodePoint(false)[1..][32..]
                }
            });
            ECDiffieHellman ecdh2 = ECDiffieHellmanCng.Create(new ECParameters
            {
                Curve = ECCurve.NamedCurves.nistP256,
                Q = new System.Security.Cryptography.ECPoint
                {
                    X = remote.EncodePoint(false)[1..][..32],
                    Y = remote.EncodePoint(false)[1..][32..]
                }
            });
            return ecdh1.DeriveKeyMaterial(ecdh2.PublicKey).Sha256();//z = r * P = r* k * G
        }
```

Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs]: Have changed.

File: tests/neo.UnitTests/Cryptography/UT_Cryptography_Helper.cs [hunk 9]
```diff
@@ -48,6 +51,48 @@ public void TestRIPEMD160()
...
+            wallet.CreateAccount();
+            wallet.CreateAccount();
+            WalletAccount account1 = wallet.GetAccounts().ToArray()[0];
+            KeyPair key1 = account1.GetKey();
+            WalletAccount account2 = wallet.GetAccounts().ToArray()[1];
+            KeyPair key2 = account2.GetKey();
+            Console.WriteLine($"Account:{1},privatekey:{key1.PrivateKey.ToHexString()},publicKey:{key1.PublicKey.ToArray().ToHexString()}");
+            Console.WriteLine($"Account:{2},privatekey:{key2.PrivateKey.ToHexString()},publicKey:{key2.PublicKey.ToArray().ToHexString()}");
+            var secret1 = Neo.Cryptography.Helper.ECDHDeriveKey(key1, key2.PublicKey);
+            var secret2 = Neo.Cryptography.Helper.ECDHDeriveKey(key1, key2.PublicKey);
```
Comment (nicolegys) [inline tests/neo.UnitTests/Cryptography/UT_Cryptography_Helper.cs:85]:
```suggestion
            var secret1 = Neo.Cryptography.Helper.ECDHDeriveKey(key1, key2.PublicKey);
            var secret2 = Neo.Cryptography.Helper.ECDHDeriveKey(key2, key1.PublicKey);
```

**Review by nicolegys — CHANGES_REQUESTED** [review]

File: src/neo/Cryptography/Helper.cs [hunk 10]
```diff
@@ -119,6 +121,63 @@ public static byte[] Sha256(this Span<byte> value)
...
+            var cipherWithTag = new byte[nonceLen + msgLen + tagLen];
+            Buffer.BlockCopy(nonce, 0, cipherWithTag, 0, nonceLen);
+            Buffer.BlockCopy(cipherBytes, 0, cipherWithTag, nonceLen, msgLen);
+            Buffer.BlockCopy(tag, 0, cipherWithTag, nonceLen + msgLen, tagLen);
+            return cipherWithTag;
+        }
+
+        public static byte[] AES256Decrypt(this byte[] encryptedData, byte[] key, byte[] associatedData = null)
+        {
+            if (key.Length != 32) throw new ArgumentException();
```
Comment (shargon) [inline src/neo/Cryptography/Helper.cs:145]: Check `encryptedData` length?

**Review by shargon — DISMISSED** [review]

**Review by erikzhang — APPROVED** [review]

File: src/neo/Cryptography/Helper.cs (hunk 10 above)
Comment (doubiliu) [reply src/neo/Cryptography/Helper.cs]: The key length,which is checked here.  It should be 32 bytes when we use  a AES gcm mode

**Review by Liaojinghui — APPROVED** [review]

**Comment by erikzhang** [comment]
Merge?

**Review by shargon — APPROVED** [review]

**Pull request merged**